### PR TITLE
 fix: don't do empty hwupload for VT

### PR DESCRIPTION
**Pull request by gnattu**
**Changes**

A hwupload without any filter to actually consume the frame would cause VideoToolbox's encoder to produce frames with artifacts. Check if the filter chain actually has a valid filter before adding hwupload.

**Issues**

### CODE REVIEW

**Review by Shadowghost — APPROVED** [review]

**Review by nyanmisaka — APPROVED** [review]
Another de-sync issue similar to when videotoolbox encodes BGRA?

Both of these functions should block until the target data is well prepared.
- hwcontext_videotoolbox.c -> [vt_transfer_data_to()](https://github.com/FFmpeg/FFmpeg/blob/b9297128f5f66bb9e6d2e755e4b82818bc1ada32/libavutil/hwcontext_videotoolbox.c#L696)
- videotoolboxenc.c -> [create_cv_pixel_buffer()](https://github.com/FFmpeg/FFmpeg/blob/b9297128f5f66bb9e6d2e755e4b82818bc1ada32/libavcodec/videotoolboxenc.c#L2450C12-L2450C34)

This is sufficient for the current use case. Best to fix it in FFmpeg in the future.

**Comment by gnattu** [comment]
> Another de-sync issue similar to when videotoolbox encodes BGRA?
> 
> Both of these functions should block until the target data is well prepared.
> 
> * hwcontext_videotoolbox.c -> [vt_transfer_data_to()](https://github.com/FFmpeg/FFmpeg/blob/b9297128f5f66bb9e6d2e755e4b82818bc1ada32/libavutil/hwcontext_videotoolbox.c#L696)
> * videotoolboxenc.c -> [create_cv_pixel_buffer()](https://github.com/FFmpeg/FFmpeg/blob/b9297128f5f66bb9e6d2e755e4b82818bc1ada32/libavcodec/videotoolboxenc.c#L2450C12-L2450C34)
> 
> This is sufficient for the current use case. Best to fix it in FFmpeg in the future.

The BGRA case is an upstream implementation issue. The upstream will initialize HEVC encoder with `kCMVideoCodecType_HEVCWithAlpha` when feed with BGRA and that encoder does not handle frame reordering properly, but `kVTCompressionPropertyKey_AllowFrameReordering` is still set to true.

This `hwhpload` one is harder to debug and I have not noticed any problematic code so far. It is a different kind of problem as the frame order is kind of OK but you will see the frames been corrupted like two or more frames are blend together.

**Comment by nyanmisaka** [comment]
> The BGRA case is an upstream implementation issue. The upstream will initialize HEVC encoder with kCMVideoCodecType_HEVCWithAlpha when feed with BGRA and that encoder does not handle frame reordering properly, but kVTCompressionPropertyKey_AllowFrameReordering is still set to true.

Is it allowed to use `kCMVideoCodecType_HEVC` with BGRA? If so, just add BGR0 to bypass encoding alpha channel.

> This hwhpload one is harder to debug and I have not noticed any problematic code so far. Just checked with more files. It is indeed a frame ordering issue. It seems like b-frame related settings is very broken.

So is disabling b-frames *fixing* it?

I haven't looked at `videotoolboxenc.c` carefully, you might know if it supports async encoding? That is, queue some frames in videotoolboxenc.c or VT runtime, and use callbacks to release these in-queue/reference frames and retrieve the encoded h26x data, which may cause artifacts if these refs frames are released prematurely.

**Comment by gnattu** [comment]
> So is disabling b-frames _fixing_ it?

No.

> I haven't looked at `videotoolboxenc.c` carefully, you might know if it supports async encoding? That is, queue some frames in videotoolboxenc.c or VT runtime, and use callbacks to release these in-queue/reference frames and retrieve the encoded h26x data, which may cause artifacts if these refs frames are released prematurely.

I set `max_ref_frames` to 0 and it does not fix the problem.

**Comment by nyanmisaka** [comment]
> Just look at the code further, when using `hwupload`, the `videotoolboxenc` will use the `hwcontext` managed `CVPixelBufferPool`, when used without, `videotoolboxenc` will create its own `CVPixelBufferPool`. Probably this is where things start to break.

But this still does not explain why the hw frames copied by `scale_vt` are not affected.

**Comment by nyanmisaka** [comment]
Apart from this,
```
// hwupload path:
CVPixelBufferPoolCreate() -> CVPixelBufferPoolCreatePixelBuffer() -> Lock -> copy -> Unlock

// encoder internal path:
VTCompressionSessionGetPixelBufferPool() -> CVPixelBufferPoolCreatePixelBuffer() -> Lock -> copy -> Unlock
```

Another difference is that the encoder internal copy path [does not use](https://github.com/FFmpeg/FFmpeg/blob/b9297128f5f66bb9e6d2e755e4b82818bc1ada32/libavcodec/videotoolboxenc.c#L2384-L2413) `av_frame_copy`.

**Comment by gnattu** [comment]
> But this still does not explain why the hw frames copied by `scale_vt` are not affected.

Yes, it's indeed quite strange. If I use my brand new `mjpeg_videotoolbox`, then this won't happen, but it does occur with `h264_videotoolbox`, `hevc_videotoolbox`, and `prores_videotoolbox`. Very hard to debug, as the code look correct everywhere.

**Comment by gnattu** [comment]
> Apart from this,
> 
> ```
> // hwupload path:
> CVPixelBufferPoolCreate() -> CVPixelBufferPoolCreatePixelBuffer() -> Lock -> copy -> Unlock
> 
> // encoder internal path:
> VTCompressionSessionGetPixelBufferPool() -> CVPixelBufferPoolCreatePixelBuffer() -> Lock -> copy -> Unlock
> ```
> 
> Another difference is that the encoder internal copy path [does not use](https://github.com/FFmpeg/FFmpeg/blob/b9297128f5f66bb9e6d2e755e4b82818bc1ada32/libavcodec/videotoolboxenc.c#L2384-L2413) `av_frame_copy`.

If this is the case then this is unfixable unless we make a not upstream-able `hwupload_videotoolbox`. I would prefer just don't hwupload without filters instead.

**Comment by nyanmisaka** [comment]
> If this is the case then this is unfixable unless we make a not upstream-able `hwupload_videotoolbox`. I would prefer just don't hwupload without filters instead.

If it's just de-sync and not distorted luma/chroma, it should have nothing to do with how to copy planes.

**Comment by gnattu** [comment]
> > If this is the case then this is unfixable unless we make a not upstream-able `hwupload_videotoolbox`. I would prefer just don't hwupload without filters instead.
> 
> If it's just de-sync and not distorted luma/chroma, it should have nothing to do with how to copy planes.

I'm not very sure:

https://github.com/jellyfin/jellyfin/assets/12995396/cdc40d51-30db-471e-a4df-1aca2727a481

It seems like desync, but if you pause you can see some frames been distorted.

**Comment by nyanmisaka** [comment]
This is still [tearing](https://en.wikipedia.org/wiki/Screen_tearing) caused by desync, no twisted chroma.

image

**Comment by gnattu** [comment]
Reported to upstream: https://trac.ffmpeg.org/ticket/10939

Let's don't think about it too much for now. There is no performance diff, and this workaround is not that ugly.

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 1]
```diff
@@ -5159,6 +5162,19 @@ public string GetHwTonemapFilter(EncodingOptions options, string hwTonemapSuffix
...
+                                subFilters.Any(f => !string.IsNullOrEmpty(f)) ||
+                                overlayFilters.Any(f => !string.IsNullOrEmpty(f));
+
+            // This is a workaround for ffmpeg's hwupload implementation
+            // For VideoToolbox encoders, a hwupload without a valid filter actually consuming its frame
+            // will cause the encoder to produce incorrect frames.
+            if (needFiltering)
+            {
+                hwuploadFilters.AddRange(mainFilters);
+                mainFilters = hwuploadFilters;
```
Comment (crobibero) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:5175]:
```suggestion
                mainFilters.Insert(0, "hwupload=derive_device=videotoolbox");
                mainFilters.Insert(0, "format=nv12|p010le|videotoolbox_vld");
```

If order doesn't matter just use `Add()`, and remove `hwuploadFilters`

Comment (nyanmisaka) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:5175]: It matters. `hwupload*` should be the first.

Comment (gnattu) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:5175]: Yes, hwupload has to be first.

Comment (crobibero) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:5175]: Ok, my suggestion is to not create a list just to maybe use it- but to just insert the required filters at the start of the list when needed

Comment (gnattu) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: Done.

**Review by crobibero — APPROVED** [review]

**Review by Bond-009 — APPROVED** [review]

**Pull request merged**